sm/plugin_sm/src/plugin.rs

```diff
@@ -34,9 +35,37 @@ pub trait Plugin {
         update: &SoftwareModuleUpdate,
         logger: &mut BufWriter<File>,
     ) -> Result<(), SoftwareError> {
-        match update {
-            SoftwareModuleUpdate::Install { module } => self.install(module, logger).await,
-            SoftwareModuleUpdate::Remove { module } => self.remove(module, logger).await,
+        match update.clone() {
+            SoftwareModuleUpdate::Install { mut module } => {
+                match &module.url {
+                    Some(url) => {
+                        let downloader = Downloader::new(&module.name, &module.version, "/tmp");
+                        let () = downloader.download(url).await.map_err(|err| {
+                            SoftwareError::DownloadError {
+                                reason: err.to_string(),
+                                url: url.url().to_string(),
+                            }
+                        })?;
+
+                        module.file_path = Some(downloader.filename().to_owned());
+
+                        let result = self.install(&module, logger).await;
+                        downloader
+                            .cleanup()
+                            .await
+                            .map_err(|err| SoftwareError::DownloadError {
+                                reason: err.to_string(),
+                                url: url.url().to_string(),
+                            })?;
+
+                        return result;
```

Comment:
To make the code clearer this might be moved into a `install_from_url` method.

Reply:
One also needs to log (in the operation log) the download operation and any download error (as we do for the install operation itself)..